src/mds/MDCache.cc

```diff
@@ -8117,8 +8118,15 @@ int MDCache::path_traverse(MDRequestRef& mdr, MDSContextFactory& cf,
       return -ENOTDIR;
     }
 
-    // walk into snapdir?
     if (path[depth].length() == 0) {
+      // this is extra slash, do nothing
+      if (!req->is_snap_op()) {
+        dout(10) << "traverse: extra slash" << dendl;
+        depth++;
+        continue;
+      }
```

Comment:
this breaks tests:

````
2019-12-06T22:08:09.286 INFO:tasks.workunit.client.0.smithi014.stderr:+ echo asdf
2019-12-06T22:08:09.286 INFO:tasks.workunit.client.0.smithi014.stderr:+ mkdir .snap/foo
2019-12-06T22:08:09.296 INFO:tasks.workunit.client.0.smithi014.stderr:+ grep asdf .snap/foo/foo
2019-12-06T22:08:09.298 INFO:tasks.workunit.client.0.smithi014.stderr:grep: .snap/foo/foo: Not a directory
2019-12-06T22:08:09.299 DEBUG:teuthology.orchestra.run:got remote process result: 2
2019-12-06T22:08:09.299 INFO:tasks.workunit:Stopping ['fs/snaps'] on client.0...
2019-12-06T22:08:09.299 INFO:teuthology.orchestra.run.smithi014:> sudo rm -rf -- /home/ubuntu/cephtest/workunits.list.client.0 /home/ubuntu/cephtest/clone.client.0
2019-12-06T22:08:09.517 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 86, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 65, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20191206.000649/qa/tasks/workunit.py", line 136, in task
    cleanup=cleanup)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20191206.000649/qa/tasks/workunit.py", line 286, in _spawn_on_all_clients
    timeout=timeout)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/parallel.py", line 87, in __exit__
    for result in self:
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/parallel.py", line 101, in __next__
    resurrect_traceback(result)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/parallel.py", line 37, in resurrect_traceback
    reraise(*exc_info)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/parallel.py", line 24, in capture_traceback
    return func(*args, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20191206.000649/qa/tasks/workunit.py", line 409, in _run_tests
    label="workunit test {workunit}".format(workunit=workunit)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 198, in run
    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 433, in run
    r.wait()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 158, in wait
    self._raise_for_status()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 180, in _raise_for_status
    node=self.hostname, label=self.label
CommandFailedError: Command failed (workunit test fs/snaps/snaptest-0.sh) on smithi014 with status 2: 'mkdir -p -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && cd -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && CEPH_CLI_TEST_DUP_COMMAND=1 CEPH_REF=95112ffcacaf8388986e900dc331d952cfa113e3 TESTDIR="/home/ubuntu/cephtest" CEPH_ARGS="--cluster ceph" CEPH_ID="0" PATH=$PATH:/usr/sbin CEPH_BASE=/home/ubuntu/cephtest/clone.client.0 CEPH_ROOT=/home/ubuntu/cephtest/clone.client.0 adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 3h /home/ubuntu/cephtest/clone.client.0/qa/workunits/fs/snaps/snaptest-0.sh'
````

From: /ceph/teuthology-archive/pdonnell-2019-12-06_03:12:34-fs-wip-pdonnell-testing-20191206.000649-distro-basic-smithi/4573530/teuthology.log

MDS log:

````
2019-12-06T22:08:09.301+0000 7f1373c68700  7 mds.0.server dispatch_client_request client_request(client.4436:50 lookup #0x10000000005//foo/foo 2019-12-06T22:08:09.302483+0000 caller_uid=1000, caller_gid=1216{6,1000,1216,10102,}) v4
2019-12-06T22:08:09.301+0000 7f1373c68700 10 mds.0.server rdlock_path_pin_ref request(client.4436:50 nref=2 cr=0x558032ebe840) #0x10000000005//foo/foo
2019-12-06T22:08:09.301+0000 7f1373c68700  7 mds.0.cache traverse: opening base ino 0x10000000005 snap head
2019-12-06T22:08:09.301+0000 7f1373c68700 10  mds.0.cache.snaprealm(0x10000000005 seq 4 0x558032ecc000) have_past_parents_open [1,head]
2019-12-06T22:08:09.301+0000 7f1373c68700 12 mds.0.cache traverse: path seg depth 0 '' snapid head
2019-12-06T22:08:09.301+0000 7f1373c68700 10 mds.0.cache traverse: extra slash
2019-12-06T22:08:09.301+0000 7f1373c68700 12 mds.0.cache traverse: path seg depth 1 'foo' snapid head
2019-12-06T22:08:09.301+0000 7f1373c68700 20 mds.0.cache.dir(0x10000000005) lookup (head, 'foo')
2019-12-06T22:08:09.301+0000 7f1373c68700 20 mds.0.cache.dir(0x10000000005)   hit -> (foo,head)
2019-12-06T22:08:09.301+0000 7f1373c68700 12 mds.0.cache traverse: path seg depth 2 'foo' snapid head
2019-12-06T22:08:09.301+0000 7f1373c68700  7 mds.0.cache traverse: [inode 0x10000000006 [4,head] /client.0/tmp/foo auth v2 pv4 ap=1 dirtyparent s=0 n(v0 1=1+0)->n(v0 rc2019-12-06T22:08:09.288066+0000 b5 1=1+0) (iauth excl) (ifile excl) (ixattr excl) (iversion lock) cr={4436=0-4194304@3} caps={4436=pAsxLsXsxFsxcrwb/pAsxXsxFxwb@2},l=4436 | ptrwaiter=1 request=0 caps=1 dirtyparent=1 dirty=1 authpin=1 0x558032e00e00] not a dir
2019-12-06T22:08:09.301+0000 7f1373c68700 10 mds.0.server FAIL on error -20
2019-12-06T22:08:09.301+0000 7f1373c68700  7 mds.0.server reply_client_request -20 ((20) Not a directory) client_request(client.4436:50 lookup #0x10000000005//foo/foo 2019-12-06T22:08:09.302483+0000 caller_uid=1000, caller_gid=1216{6,1000,1216,10102,}) v4
2019-12-06T22:08:09.301+0000 7f1373c68700 10 mds.0.server apply_allocated_inos 0x0 / [] / 0x0
2019-12-06T22:08:09.301+0000 7f1373c68700 20 mds.0.server lat 0.000064
2019-12-06T22:08:09.301+0000 7f1373c68700 10 mds.0.5 send_message_client client.4436 172.21.15.14:0/3727713856 client_reply(???:50 = -20 (20) Not a directory) v1
2019-12-06T22:08:09.301+0000 7f1373c68700  1 -- [v2:172.21.15.14:6832/3137234131,v1:172.21.15.14:6833/3137234131] --> 172.21.15.14:0/3727713856 -- client_reply(???:50 = -20 (20) Not a directory) v1 -- 0x558032cd9080 con 0x558031ffc800
````

From: /ceph/teuthology-archive/pdonnell-2019-12-06_03:12:34-fs-wip-pdonnell-testing-20191206.000649-distro-basic-smithi/4573530/remote/smithi014/log/ceph-mds.a.log.gz

I'm not sure this fix is safe. ping @ukernel

Reply:
When looking up the .snap/foo/foo/, currently only the .snap/ will set as CEPH_SNAPDIR, and the first foo/ will returned as the CEPH_NOSNAP. So when looking up the second foo/, the op will be CEPH_MDS_OP_LOOKUP instead of CEPH_MDS_OP_LOOKUPSNAP.

Reply:
```
diff --git a/src/client/Client.cc b/src/client/Client.cc
index e042539c3b..12a539e901 100644
--- a/src/client/Client.cc
+++ b/src/client/Client.cc
@@ -6251,7 +6251,7 @@ void Client::renew_caps(MetaSession *session)
 int Client::_do_lookup(Inode *dir, const string& name, int mask,
                       InodeRef *target, const UserPerm& perms)
 {
-  int op = dir->snapid == CEPH_SNAPDIR ? CEPH_MDS_OP_LOOKUPSNAP : CEPH_MDS_OP_LOOKUP;
+  int op = dir->snapid != CEPH_NOSNAP ? CEPH_MDS_OP_LOOKUPSNAP : CEPH_MDS_OP_LOOKUP;
   MetaRequest *req = new MetaRequest(op);
   filepath path;
   dir->make_nosnap_relative_path(path);

```

Maybe the above could fix it and I will test it later.